### DISCUSSION TITLE
Where is the operating system implementation defined in the source ?

### DISCUSSION THREAD

**Discussion by SeleDreams** [dotnet/runtime#72391] [Q&A] [answered]
Hi, I was thinking of checking the operating system specific implementation in the source, hoping to find folders for each specific OS defining the APIs available etc to check the possibility of porting to some platforms, however I am not too sure where to look

do you knnow where the operating system APIs are defined ?

**Comment by gfoidl** [comment, marked as answer]
Basically in https://github.com/dotnet/runtime/tree/main/src/native
OS specific code is mostly `#if`defed or included via CMake.

**Reply by SeleDreams** [reply]
Ah I see, that's kinda unfortunate.
I tend to prefer the way game engines like godot handle platform specific code, they make abstract base classes for stuff depending on OS specific APIs and then those get implemented in their own directories and source files, it feels cleaner as the platform specific code ends up completely independent from the rest

**Reply by huoyaoyuan** [reply]
Some Unix APIs have common C interface, but no stable ABI that's required by P/Invoke. We thus implement a thin layer in C that stabilize the ABI.
